### PR TITLE
data encryption

### PR DESCRIPTION
**Pull request by Crashdown59**
oio/api/backblaze : adding cryptography
oio/common : adding some cryptography tools
oio/cli : adding a new openio command : openio generate encryption key

Some notes : The cryptography is actually made with the requests library with the Fernet method.
To see more details about Fernet, see this page : https://github.com/fernet/spec/blob/master/Spec.md

### CODE REVIEW

File: oio/common/cryptography_tools.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    # content : if fd_out is None, this dictionnary entry returns
+    # the ciphered_content
+    # WARNING : every fd_out writing is in the responsability of the programmer
+    # if you must seek, truncate or others operations, the responsability is
+    # your own!
+    # Hooks available :
+    # on_ciphered_data(data) -> new_ciphered_data
+    # treatment to make with the data encrypted
+    # on_write(data) -> None
+    # treatment to make with the data written or sended
```
Comment (fvennetier) [inline oio/common/cryptography_tools.py]: Could you format this as a python docstring?

File: oio/common/cryptography_tools.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    # this function can be used to passing to the new cryptography system
+    def from_crypt_function_to_actual_crypt(self, decrypt, cipher):
+        decrypt_content = decrypt(cipher)
+        return self.encrypt(decrypt_content)
+
+    def _generate_dict(self, over, bytes_read, ciphered_bytes, content=None):
+        return {'over': over,
+                'bytes_read': bytes_read,
+                'ciphered_bytes': ciphered_bytes,
+                'content': content}
```
Comment (fvennetier) [inline oio/common/cryptography_tools.py]: What does this function do? Its name does not tell it.

File: oio/common/cryptography_tools.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+            # False is just the end of the file
+            yield self._generate_dict(True, len(data),
+                                      file_transferred, content)
+
+    # Takes a fd and decrypt the content of ciphered content in the descriptor
+    # yielding every tokens available
+    # one hook available :
+    # buffer_to_tokens(stream) : generator
+    # -> isolate the fd into multiple tokens to be decrypted
+    def decrypt_from_buffer(self, stream, begin, end, hooks={}):
```
Comment (fvennetier) [inline oio/common/cryptography_tools.py]: The name of this function is misleading. It works with a stream, not a buffer.

File: oio/common/cryptography_tools.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+                actual += len(content)
+                token_nb += 1
+            except CryptographyToolsException as e:
+                raise OioException('Cryptography exception: %s' % (str(e)))
+
+    def get_token_size(self, data_size):
+        remains_size = (data_size * 8) % 128
+        encrypt_entire_size = 128 - remains_size + (data_size * 8)
+        overhead_size = 128 + 8 + 64 + 256
+        return (overhead_size + encrypt_entire_size) / 8
```
Comment (fvennetier) [inline oio/common/cryptography_tools.py]: https://youtu.be/0p_1QSUsbsM?t=10

File: oio/common/cryptography_tools.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+
+    def is_empty_key(self):
+        return False if self.crypt else True
```
Comment (fvennetier) [inline oio/common/cryptography_tools.py]: The key is not the point here. You don't want to know if there is a key, you want to know if the instance does something to the data. A name like `is_noop` would be better.

File: oio/api/backblaze.py [hunk 6]
```diff
@@ -194,17 +158,41 @@ def _stream_big_chunks(self, source, conn, temp):
...
+            # we recover the information generator
+            gen = self.backblaze_info['encryption'].read_and_encrypt(
+                source, size, Backblaze.BACKBLAZE_MAX_CHUNK_SIZE, temp,
+                hooks)
+            # we recover the first part
+            result = next(gen)
+            over = result['over']
+            # if the is_big parameter is True, it means that the content is
+            # greater than b2_max_chunk_size
+            if not over:
```
Comment (fvennetier) [inline oio/api/backblaze.py]: I think you want the opposite.

File: oio/common/cryptography_tools.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+
+    def _generate_dict(self, over, bytes_read, ciphered_bytes, content=None):
+        """
+        generate a dictionnary with some informations,
+        see the function read_and encrypt for more details
+        """
+        return {'over': over,
+                'bytes_read': bytes_read,
+                'ciphered_bytes': ciphered_bytes,
+                'content': content}
```
Comment (fvennetier) [inline oio/common/cryptography_tools.py]: To generate a dictionary, you call `dict()`. The purpose of this function is not to _generate a dictionary_, so please rename it.

**Comment by jfsmig** [comment]
:man_shrugging:

**Pull request closed without merging**